### DISCUSSION TITLE
Deepgram doesn't close websocket on finish()?

### DISCUSSION THREAD

**Discussion by sh-ankush92** [deepgram/community#619] [General help]
### Which Deepgram product are you using?

Deepgram SDKs

### Details

Facing a persistent issue in Live Transcription from a week. I have tried everything in my power to fix it but unsuccessful so far. Need help! 

The issue is very similar to this thread: https://github.com/orgs/deepgram/discussions/597 

My environment is this: 
1. Client records from microphone on browser using MediaRecorder API. Encoding: webm/opus
2. Client sends this to my websocket server in backend 
3. As soon as audio packets starting coming to my backend, server opens a Deepgram socket using Javascript SDKs -> client.listen.live()
4. Client stops sending packets, server closes connection to Deepgram socket using finish() 

Issues that I am facing -> Deepgram transcript is unstable
1. Doesn't work on second run
This setup generally works fine only at the first instance (Start Recording -> Stop Recording) on browser. I get transcripts one after the other as expected. But, the second time I try to get transcript (Start Recording -> Stop Recording), deepgram disconnects after sending 3-4 packets (~0.01s) and it doesn't give me ANY transcript from then.. To get it to work again, I have to restart my backend server which is annoying. 

2. Doesn't work intermittently in a session
When having multiple turns in the same conversation, deepgram suddenly disconnects and doesn't respond henceforth. 
 
What I have done so far: 
1. I need to send headers for webm/opus everytime a deepgram socket is opened -> tried every permutation/combination here.. made sure mic is properly closed and restarted and inspected headers to the last byte to make sure the headers are going everytime deepgram socket is opened -> DOESN'T WORK!!! 
2. Stored the first ever packet and forcibly sending it as first packet for every socket OPEN like suggested here (https://github.com/orgs/deepgram/discussions/597) -> DOESN'T WORK
4. Then I suspected if websocket is being closed properly.. finish() didn't seem to work, so I explicitly did live._socket.close() -> DOESN'T WORK

Can someone please help me. We need to fix this immediately!

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Reply by davidvonthenen** [reply]
Can you share your connection options on the live client?
Also did you make sure the match your audio stream type and the `encoding` value on the live client options?

**Reply by sh-ankush92** [reply]
@dvonthenen, here are the connection options that I am using: 
model: nova-2-conversationalai,
      endpointing: 500, 
      interim_results: true,
      utterance_end_ms: 1000,
      encoding: "webm-opus",
      sample_rate: "48000",
      channels: 2,
      
      I have tried multiple variations of encoding, sampling, channels and even removing them completely. 
      The issue is this: 
1. for first run, it works 8/10 times without any issue.. the other 2 times, I get deepgram close event 
2. for first run, 1/10 times it can close midway in the conversation
3. second run onwards, it doesn't work at all even though I ensure the webm headers are resent (mic is reinitialized) and I have personally verified the header bytes that is going in the first packet. 

I have to restart my client (browser) or restart my backend server to make it "first run" again

**Reply by sh-ankush92** [reply]
To provide additional information that can help in the debug, in the second run, I am writing the file to disk and trying to open in VLC media player.. it doesn't open although the headers are exactly what they are supposed to be for audio/webm.. I may be pushing incorrect data to deepgram on second run but I am not able to figure out what since the headers are all correct..

**Reply by sh-ankush92** [reply]
Header bytes in second run (not working): 
![image](https://github.com/deepgram/community/assets/150580268/46bafd68-3b5d-4314-855f-e6ae24eaf14e)


Header bytes in first run (working):
![image](https://github.com/deepgram/community/assets/150580268/077371fa-991a-4b24-b545-14ccb17d7771)

**Comment by davidvonthenen** [comment]
hi @ankushdigitar 

you `encoding` value is not valid. take a look at the documentation:
https://developers.deepgram.com/docs/encoding

**Reply by sh-ankush92** [reply]
@dvonthenen , from the above documentation
"Encoding is required when raw, headerless audio packets are sent to the streaming service. If containerized audio packets are sent to the streaming service, this feature should not be used" 

However, I am sending containerized "webm" audio packets, I tried removing encoding, sampling rate etc., but it didn't work..

**Reply by davidvonthenen** [reply]
Sorry, I mean `webm-opus` is not a valid value in the list.

**Reply by davidvonthenen** [reply]
There was a release that was just posted that might address the `finish()` problems you might be facing.
https://github.com/deepgram/deepgram-python-sdk/releases/tag/v3.1.8
